src/kili/gateways/kili_api_gateway/queries.py

```diff
@@ -0,0 +1,164 @@
+                    elements = self._graphql_client.execute(query, payload)["data"]
+
+                    if elements is None or len(elements) == 0:
+                        break
+
+                    if post_call_function is not None:
+                        elements = post_call_function(elements)
+
+                    if isinstance(elements, Dict):
+                        yield elements
+                        break
+
+                    yield from elements
+
+                    count_elements_retrieved += len(elements)
+                    pbar.update(len(elements))
+
+                    if len(elements) < first:
+                        break
+
+
+def get_number_of_elements_to_query(
+    graphql_client: GraphQLClient,
+    count_query: str,
+    where: AbstractQueryWhere,
+    options: QueryOptions,
+) -> int:
+    """Give the total number of elements to query for one query that will be paginated.
+
+    It uses both the argument first given by the user
+    and the total number of available objects obtained with a graphQL count query
+
+    Args:
+        graphql_client: The graphQL client
+        where: the where payload to the count query,
+        options: the query options with skip and first
+        count_query: the count query
+
+    Returns:
+        The number of elements to query
+    """
+    first = options.first
+    skip = options.skip
+    payload = {"where": where.build_gql_where()}
+    count_result = graphql_client.execute(count_query, payload)
+    nb_elements = count_result["data"]
+    nb_elements_queried = max(nb_elements - skip, 0)
+    if first is None:
+        return nb_elements_queried
+    return min(nb_elements_queried, first)
+
+
+@typechecked
+def fragment_builder(fields: List[str]):
+    """Builds a GraphQL fragment for a list of fields to query.
+
+    Args:
+        fields: The list of fields to query
+    """
+    fragment = ""
+
+    # split a field and its subfields (e.g. "roles.user.id" -> ["roles", "user.id"])
+    subfields = [field.split(".", 1) for field in fields if "." in field]
+
+    if subfields:
+        # get the root fields (e.g. "roles" in "roles.user.id")
+        root_fields = {subfield[0] for subfield in subfields}
+        for root_field in root_fields:
+            # get the subfields of the root field (e.g. "user.id" in "roles.user.id")
+            fields_subquery = [subfield[1] for subfield in subfields if subfield[0] == root_field]
+            # build the subquery fragment (e.g. "user{id}" in "roles{user{id}}")
+            new_fragment = fragment_builder(fields_subquery)
+            # add the subquery to the fragment
+            fragment += f" {root_field}{{{new_fragment}}}"
+
+        # remove the fields that have been queried in subqueries (e.g. "roles.user.id")
+        fields = [field for field in fields if "." not in field]
+
+    for field in fields:
+        fragment += f" {field}"
+
+    return fragment
```

Comment:
you moved fragment_builder here but didn't remove it from `class GraphQLQuery(ABC)`, we should avoid duplicating code

this method is can be removed from the class, it doesn't really need the `self` anyway

Reply:
good idea